### ISSUE TITLE
Textarea's cursor jumps when changing its value through reflux

### ISSUE THREAD

**Issue by raine** [reflux/refluxjs#43]
Hello,

I'm experiencing an issue where if I update a textarea's value through the `Action → Store → View` flow, with the cursor _not_ at the end, the cursor will jump to the end. The same doesn't happen if I use state internally in the component to keep track of the current value, like in [this example](http://facebook.github.io/react/docs/forms.html#controlled-components).

Does this by an off chance have to do something with this?

> Action and store callbacks are not executed in a single synchronous iteration

The code is something like this:

```
render: function() {
  return (
    <textarea
      value={this.props.value}
      onChange={this.handleChange}
    />
  );
},

handleChange: function(ev) {
  Actions.updateMessage(ev.target.value);
}
```

Also noticed this one, https://github.com/facebook/react/issues/1698

**Comment by damianmr** [comment]
Got into the same exact problem here. 

I have one Store that is listened by only one element, which happens to be the root element of my app. If want an input to update the state of the store by calling an action the store is listening to, the input refreshes its state to whatever the store triggers (because, you know, the input is descendant of the root node and got some props from it). The thing is that if the caret is in the middle of the text it moves to the end of the text when React updates the value of the input.

In my case, I can just change the event handling so that the input updates its state in the onBlur event, but I can certainly see your point @ericclemmons where it would be very hard to do some kind of transformation during the onChange.


I am not sure this is a specific Reflux issue, but I would like to know what workarounds other people have found.

**Comment by HurricaneJames** [comment]
Hey, I ran into this problem yesterday and found this thread. I try not to necro old threads, especially closed ones, but this thread recently had a post (less than 2 weeks ago). Since this seems to be a common problem, I created a LazyInput component based on the optimistic update approach that we use on our team. The component is up on [GitHub](https://github.com/HurricaneJames/lazy-input) and a npm package on [npmjs](https://www.npmjs.com/package/lazy-input).

**Comment by ericclemmons** [comment]
Thank you for necro'ing this thread.  It's a big problem (still!) and I hope LazyInput solves it.  I'll be testing soon...

**Comment by AlexJozwicki** [comment]
Hope I haven't missed the point here but, inputs in general should solely be used with valueLink/checkedLink.
Even without Reflux in the loop and doing a simple setState will make the caret jump, just as in the fiddle in referenced in facebook/react#1698.
Same fiddle with valueLink works like a charm though.

We did tag in our projects the actions being triggered by the valueLink as sync, which makes sense in the case of an input.
Don't hesitate to create your own link object and not the one provided by the React mixin; you can them trigger a second action async (debounced?) if needed.

**Comment by HurricaneJames** [comment]
I do not use valueLink/checkedLink, but even if I did, they are just syntactical sugar. Controlled components do not, in general have this problem because the `input` component forces a render with the updated state at just the right moment. That is not an option with Reflux.

lazy-input fixed the problem, at first. However, as the form grew (up to a couple hundred fields so far) it started fraying at the seams. Specifically, it had real issues with typing fast. The input would start overwriting itself. After a couple iterations, I finally found something that works.  Basically, lazy-input was not lazy enough. In addition to delaying updates until new props are presented, lazy-input only updates after a set period of time after any change that originated from the input.

Undoubtedly, somebody will find an edge case that breaks the input, but so far it solves both the cursor jumping and the performance issues.

[lazy-input (github)](https://github.com/HurricaneJames/lazy-input)
[lazy-input (npmjs)](https://www.npmjs.com/package/lazy-input)

**Comment by AlexJozwicki** [comment]
In this case valueLink will indeed do nothing to help you.
We had the exact same issue in our project of input starting to eat itself.

We tagged all the actions triggered by requestChange/onChange to sync in Reflux, and splitting those who needed into two: one for just updating the text backing the input as sync, and another lunching some longer task in async such as form validation/search/server queries.

I really do not think that the actions triggered by the input and actively modifying the data backing it should be done asynchronously.

**Comment by HurricaneJames** [comment]
I tried tagging the actions as sync, it didn't help. I also async anything long like validation or search. Nothing helped, but I'm also working on an insanely complex form with hundreds of fields. Re-render, even with smart use of shouldComponentUpdate, takes too much time for interactive input. A semi-controlled/lazy input is the only solution I found that seems to work, especially once we hit mobile.

There are still delays, and sometimes the rest of the page continues to update for a few seconds after typing stops, but the active input stays perfectly in sync with the input.

**Comment by prajwalit** [comment]
This is how I fixed it -
```js
getInitialState: function () {
  return {val: this.props.value}
},

render: function() {
  return (
    <textarea
      value={this.state.val}
      onChange={this.handleChange}
    />
  );
},

handleChange: function(ev) {
  this.setState ({val: ev.target.value});
  Actions.updateMessage(ev.target.value);
},

componentWillReceiveProps: function (nextProps) {
  if (this.state.val !== nextProps.value) {
    this.setState ({val: nextProps.value});
  }
}
```

[Ref](https://facebook.github.io/react/docs/component-specs.html#updating-componentwillreceiveprops)

What do you guys think?

**Comment by srph** [comment]
@prajwalit Do you have a working demo?

**Comment by HurricaneJames** [comment]
@prajwalit that is basically how [lazy-input](https://github.com/HurricaneJames/lazy-input) works, except it is also useful to add a timer to only update the textarea only after a set amount of time. Otherwise anything that slows down the sync will cause the store to come back a few ms later and change the value, seemingly randomly deleting characters as you type.

**Comment by prajwalit** [comment]
@srph 
Here's the problem: http://jsfiddle.net/prajwalit/Ln3261n2/1/
And the solution: http://jsfiddle.net/prajwalit/Ln3261n2/2/

This doesn't demo what happens when you change props by some action outside the components. But `componentWillReceiveProps` will take care of any changes to the inputValue outside the component as it'll update the state if prop goes out of sync.

@HurricaneJames Oh, I didn't face the slowness issue. I will test a bit more. Thanks! What was slowing down the sync in your case?

**Comment by HurricaneJames** [comment]
@prajwalit The problem comes up when you have more than just a couple components on the page (in my cause there are probably over 1k components by now). Even with `shouldComponentUpdate` pruning, it can take a 50-150 ms to process all of those. That said, I managed to reproduce it in your solution, but it is harder because it is a very simple case.

First type a few characters. Then put the cursor in the middle of those characters. Then run your hand across the keyboard as fast as possible (I recommend dragging it back and forth across the keys like a crazy person). At some point you will probably notice that you are typing at the end of the line. That lets you know that you exceeded the speed at which input could be processed. Because the input was so random, you probably won't notice, but you probably deleted some characters too.

To prove it, I added some latency to your solution on jsfiddle, [http://jsfiddle.net/Ln3261n2/3/](http://jsfiddle.net/Ln3261n2/3/). At 100ms of latency, it is pretty bad. I could feel it, consistently all the way down to 50ms latency on my machine, and it is a fairly powerful macbook pro. Imagine what would happen on a slower computer.

**Comment by prajwalit** [comment]
@HurricaneJames Ah, I see. That's a really nice catch. Turns ugly when you press and hold delete button.
Can I improve upon my solution? Though I feel like it's getting closer and closer to the lazy-input.
http://jsfiddle.net/prajwalit/Ln3261n2/4/

**Comment by HurricaneJames** [comment]
Don't forget to have `handleChange` reset your timer at the same too, otherwise you can get some weird race conditions that cause input to disappear. Of course once you do that you have basically re-created lazy-input.

**Comment by prajwalit** [comment]
Yep. Thanks @HurricaneJames!

**Comment by benglass** [comment]
I know this issue is closed but we just ran into this problem which was solved by adding sync: true to the actions and it might be nice to update the docs to explain this common use case of sync (form input values managed through actions) because I had no idea what it did before we ran into this problem

**Comment by lmtorres** [comment]
@benglass Can you expand a little bit on this? I've been running into this problem and went the timer route, but it definitely feels 'hacky'.

**Comment by LoicUV** [comment]
Have you guys tried to use `defaultValue` instead of `value` ? It fixes the problem for me.

**Comment by benglass** [comment]
@lmtorres I believe the issue is that without sync: true, the action is not actually fired immediately because reflux uses nextTick to queue it up. This slight delay makes it possible that the user can continue to typing before the change makes it to the store and back and this is why you cannot use sync: false actions for managing the value of form fields.

@LoicUV You are correct that defaultValue works around the issue but that is just because it turns the field into an un-managed input. This bug report is NOT about un-managed inputs. If you want to have a managed input (with value and onChange set) where you are dispatching an action to update the field value in a store and then propagating the change back down to the input, you need to use sync: true to ensure the action is triggered immediately instead of deferred to the next tick.

**Comment by LoicUV** [comment]
Thx for the clarification @benglass
